File: snowcap/api/lua/snowcap/layer.lua
-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at https://mozilla.org/MPL/2.0/.

local log = require("snowcap.log")
local client = require("snowcap.grpc.client").client

local widget = require("snowcap.widget")
local widget_signal = require("snowcap.widget.signal")

---@class snowcap.layer
local layer = {}

local layer_handle = {}

---@class snowcap.layer.LayerHandle
---@field id integer
---@field private _update fun(msg:any)
---@field private _operate fun(oper: snowcap.widget.operation.Operation)
local LayerHandle = {}

---Convert a LayerHandle into a Popup's ParentHandle
---@return snowcap.popup.ParentHandle
function LayerHandle:as_parent()
    return require("snowcap.popup").parent.Layer(self)
end

---@param id integer
---@param update fun(msg: any?)
---@param operate fun(oper: snowcap.widget.operation.Operation)
---@return snowcap.layer.LayerHandle
function layer_handle.new(id, update, operate)
    ---@type snowcap.layer.LayerHandle
    local self = {
        id = id,
        _update = update,
        _operate = operate,
    }
    setmetatable(self, { __index = LayerHandle })
    return self
end

---@enum snowcap.layer.Anchor
local anchor = {
    TOP = 1,
    BOTTOM = 2,
    LEFT = 3,
    RIGHT = 4,
    TOP_LEFT = 5,
    TOP_RIGHT = 6,
    BOTTOM_LEFT = 7,
    BOTTOM_RIGHT = 8,
}

---@enum snowcap.layer.KeyboardInteractivity
local keyboard_interactivity = {
    NONE = 1,
    ON_DEMAND = 2,
    EXCLUSIVE = 3,
}

---@enum snowcap.layer.ZLayer
local zlayer = {
    BACKGROUND = 1,
    BOTTOM = 2,
    TOP = 3,
    OVERLAY = 4,
}

---@package
---@enum snowcap.layer.FocusEvent
local focus_event = {
    GAINED = 1,
    LOST = 2,
}

---@alias snowcap.layer.ExclusiveZone
---| integer
---| "respect"
---| "ignore"

---@param zone snowcap.layer.ExclusiveZone
---@return integer
local function exclusive_zone_to_api(zone)
    if type(zone) == "number" then
        return zone
    end

    if zone == "respect" then
        return 0
    end

    return -1
end

---@class snowcap.layer.LayerArgs
---@field program snowcap.widget.Program
---@field anchor snowcap.layer.Anchor?
---@field keyboard_interactivity snowcap.layer.KeyboardInteractivity
---@field exclusive_zone snowcap.layer.ExclusiveZone
---@field layer snowcap.layer.ZLayer

---@param args snowcap.layer.LayerArgs
---@return snowcap.layer.LayerHandle|nil handle A handle to the layer surface, or nil if an error occurred.
function layer.new_widget(args)
    ---@type table<integer, any>
    local callbacks = {}

    local widget_def = args.program:view() or widget.row({ children = {} })

    widget._traverse_widget_tree(widget_def, callbacks, widget._collect_callbacks)

    ---@type snowcap.layer.v1.NewLayerRequest
    local request = {
        layer = args.layer,
        exclusive_zone = exclusive_zone_to_api(args.exclusive_zone),
        anchor = args.anchor,
        keyboard_interactivity = args.keyboard_interactivity,
        widget_def = widget.widget_def_into_api(widget_def),
    }

    local response, err = client:snowcap_layer_v1_LayerService_NewLayer(request)

    if err then
        log.error(err)
        return nil
    end

    assert(response)

    if not response.layer_id then
        log.error("no layer_id received")
        return nil
    end

    local layer_id = response.layer_id or 0

    local sender, receiver = require("snowcap.util.channel").spsc()

    ---@type fun(msg: any?)
    local update_on_msg = function(msg)
        if msg ~= nil then
            sender:send({
                message = msg
            })
        else
            sender:send({
                redraw = {}
            })
        end
    end

    ---@type fun(oper: snowcap.widget.operation.Operation)
    local operate_surface = function(oper)
        sender:send({
            operation = oper
        })
    end

    local handle = layer_handle.new(layer_id, update_on_msg, operate_surface)

    ---@type fun(): snowcap.signal.HandlerPolicy
    local close_surface = function()
        handle:close()

        return require("snowcap.signal").HandlerPolicy.Discard
    end

    args.program:connect(widget_signal.redraw_needed, update_on_msg)
    args.program:connect(widget_signal.send_message, update_on_msg)
    args.program:connect(widget_signal.operation, operate_surface)
    args.program:connect(widget_signal.request_close, close_surface)

    args.program:event({
        created = widget.SurfaceHandle.from_layer_handle(handle),
    })

    err = client:snowcap_layer_v1_LayerService_GetLayerEvents({
        layer_id = layer_id,
    }, function(response) ---@diagnostic disable-line:redefined-local
        sender:send({
            layer_events = response.layer_events or {}
        })
    end)

    err = client:snowcap_widget_v1_WidgetService_GetWidgetEvents({
        layer_id = layer_id,
    }, function(response) ---@diagnostic disable-line:redefined-local
        sender:send({
            widget_events = response.widget_events or {}
        })
    end)

    client.loop:wrap(function()
        local pending_operations = {}

        local msg = receiver:recv()
        while msg do
            local update_view = false;

            if msg.widget_events then
                for _, event in ipairs(msg.widget_events) do
                    ---@diagnostic disable-next-line:invisible
                    local message = widget._message_from_event(callbacks, event)

                    if message then
                        local ok, update_err = pcall(function()
                            args.program:update(message)
                        end)
                        if not ok then
                            log.error(update_err)
                        end
                    end
                end
                update_view = true
            elseif msg.message then
                args.program:update(msg.message)
            elseif msg.operation then
                table.insert(pending_operations, msg.operation)
            elseif msg.layer_events then
                for _, layer_event in ipairs(msg.layer_events) do
                    if layer_event.closing ~= nil then
                        goto main_loop_break
                    end

                    local focus = layer_event.focus --[[@as snowcap.layer.FocusEvent]]
                    ---@type snowcap.widget.SurfaceEvent?
                    local event = nil

                    if focus == focus_event.GAINED then
                        event = {
                            focus_gained = {},
                        }
                    elseif focus == focus_event.LOST then
                        event = {
                            focus_lost = {},
                        }
                    end

                    if event then
                        args.program:event(event)
                    end
                end
            end

            if not update_view then
                ---@diagnostic disable-next-line: redefined-local
                local _, err = client:snowcap_layer_v1_LayerService_RequestView({
                    layer_id = layer_id,
                })

                if err then
                    log.error(err)
                end
            else
                ---@diagnostic disable-next-line:redefined-local
                local widget_def = args.program:view() or widget.row({ children = {} })
                callbacks = {}

                widget._traverse_widget_tree(widget_def, callbacks, widget._collect_callbacks)

                ---@diagnostic disable-next-line:redefined-local
                local _, err = client:snowcap_layer_v1_LayerService_UpdateLayer({
                    layer_id = layer_id,
                    widget_def = widget.widget_def_into_api(widget_def),
                })

                if err then
                    log.error(err)
                end

                for _, oper in pairs(pending_operations) do
                    ---@diagnostic disable-next-line:redefined-local
                    local _, err = client:snowcap_layer_v1_LayerService_OperateLayer({
                        layer_id = layer_id,
                        operation = require("snowcap.widget.operation")._to_api(oper), ---@diagnostic disable-line: invisible
                    })

                    if err then
                        log.error(err)
                    end
                end
                pending_operations = {}
            end

            msg = receiver:recv()
        end
        ::main_loop_break::

        args.program:event({
            closing = {},
        })
    end)

    return handle
end

---Do something when a key event is received.
---@param on_event fun(handle: snowcap.layer.LayerHandle, event: snowcap.input.KeyEvent)
function LayerHandle:on_key_event(on_event)
    local err = client:snowcap_input_v1_InputService_KeyboardKey(
        { layer_id = self.id },
        function(response)
            ---@cast response snowcap.input.v1.KeyboardKeyResponse

            local mods = response.modifiers or {}
            mods.shift = mods.shift or false
            mods.ctrl = mods.ctrl or false
            mods.alt = mods.alt or false
            mods.super = mods.super or false

            ---@cast mods snowcap.input.Modifiers

            ---@type snowcap.input.KeyEvent
            local event = {
                key = response.key or 0,
                mods = mods,
                pressed = response.pressed,
                captured = response.captured,
                text = response.text,
            }

            on_event(self, event)
        end
    )

    if err then
        log.error(err)
    end
end

---@param on_press fun(mods: snowcap.input.Modifiers, key: snowcap.Key)
function LayerHandle:on_key_press(on_press)
    self:on_key_event(function(_, event)
        if not event.pressed or event.captured then
            return
        end

        on_press(event.mods, event.key)
    end)
end

---@class snowcap.layer.LayerUpdateArgs
---@field anchor? snowcap.layer.Anchor
---@field keyboard_interactivity? snowcap.layer.KeyboardInteractivity
---@field exclusive_zone? snowcap.layer.ExclusiveZone
---@field layer? snowcap.layer.ZLayer

---Update this layer's attributes.
---@param args snowcap.layer.LayerUpdateArgs
---@return boolean True if the operation succeed.
function LayerHandle:update(args)
    local exclusive_zone = args.exclusive_zone and exclusive_zone_to_api(args.exclusive_zone) or nil

    local _, err = client:snowcap_layer_v1_LayerService_UpdateLayer({
        layer_id = self.id,
        anchor = args.anchor,
        keyboard_interactivity = args.keyboard_interactivity,
        exclusive_zone = exclusive_zone,
        layer = args.layer,
    })

    if err then
        log.error(err)
    end

    return err == nil
end

function LayerHandle:close()
    local _, err = client:snowcap_layer_v1_LayerService_Close({ layer_id = self.id })

    if err then
        log.error(err)
    end
end

function LayerHandle:send_message(message)
    self._update(message)
end

---Sends an `Operation` to this layer.
---@param operation snowcap.widget.operation.Operation
function LayerHandle:operate(operation)
    self._operate(operation)
end

layer.anchor = anchor
layer.keyboard_interactivity = keyboard_interactivity
layer.zlayer = zlayer

return layer

File: snowcap/api/rust/src/surface/layer.rs
//! Support for layer surface widgets using `wlr-layer-shell`.

use std::{collections::HashMap, num::NonZeroU32};

use snowcap_api_defs::snowcap::{
    input::v1::{KeyboardKeyRequest, keyboard_key_request::Target},
    layer::{
        self,
        v1::{CloseRequest, NewLayerRequest, OperateLayerRequest, UpdateLayerRequest, ViewRequest},
    },
    widget::v1::{GetWidgetEventsRequest, get_widget_events_request},
};
use tokio::sync::mpsc::UnboundedSender;
use tokio_stream::StreamExt;
use tracing::error;
use xkbcommon::xkb::Keysym;

use crate::{
    BlockOnTokio,
    client::Client,
    input::{KeyEvent, Modifiers},
    popup::{self, AsParent},
    widget::{self, Program, WidgetDef, WidgetId, WidgetMessage, operation, signal},
};

// TODO: change to bitflag
/// An anchor for a layer surface.
#[allow(missing_docs)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum Anchor {
    Top,
    Bottom,
    Left,
    Right,
    TopLeft,
    TopRight,
    BottomLeft,
    BottomRight,
}

impl From<Anchor> for layer::v1::Anchor {
    fn from(value: Anchor) -> Self {
        match value {
            Anchor::Top => layer::v1::Anchor::Top,
            Anchor::Bottom => layer::v1::Anchor::Bottom,
            Anchor::Left => layer::v1::Anchor::Left,
            Anchor::Right => layer::v1::Anchor::Right,
            Anchor::TopLeft => layer::v1::Anchor::TopLeft,
            Anchor::TopRight => layer::v1::Anchor::TopRight,
            Anchor::BottomLeft => layer::v1::Anchor::BottomLeft,
            Anchor::BottomRight => layer::v1::Anchor::BottomRight,
        }
    }
}

/// Layer surface keyboard interactivity.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum KeyboardInteractivity {
    /// This layer surface cannot get keyboard focus.
    None,
    /// This layer surface can get keyboard focus through the compositor's implementation.
    OnDemand,
    /// This layer surface will take exclusive keyboard focus.
    Exclusive,
}

impl From<KeyboardInteractivity> for layer::v1::KeyboardInteractivity {
    fn from(value: KeyboardInteractivity) -> Self {
        match value {
            KeyboardInteractivity::None => layer::v1::KeyboardInteractivity::None,
            KeyboardInteractivity::OnDemand => layer::v1::KeyboardInteractivity::OnDemand,
            KeyboardInteractivity::Exclusive => layer::v1::KeyboardInteractivity::Exclusive,
        }
    }
}

/// Layer surface behavior for exclusive zones.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum ExclusiveZone {
    /// This layer surface requests an exclusive zone of the given size.
    Exclusive(NonZeroU32),
    /// The layer surface does not request an exclusive zone but wants to be
    /// positioned respecting any active exclusive zones.
    Respect,
    /// The layer surface does not request an exclusive zone and wants to be
    /// positioned ignoring any active exclusive zones.
    Ignore,
}

impl From<ExclusiveZone> for i32 {
    fn from(value: ExclusiveZone) -> Self {
        match value {
            ExclusiveZone::Exclusive(size) => size.get() as i32,
            ExclusiveZone::Respect => 0,
            ExclusiveZone::Ignore => -1,
        }
    }
}

/// The layer on which a layer surface will be drawn.
#[allow(missing_docs)]
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum ZLayer {
    Background,
    Bottom,
    Top,
    Overlay,
}

impl From<ZLayer> for layer::v1::Layer {
    fn from(value: ZLayer) -> Self {
        match value {
            ZLayer::Background => Self::Background,
            ZLayer::Bottom => Self::Bottom,
            ZLayer::Top => Self::Top,
            ZLayer::Overlay => Self::Overlay,
        }
    }
}

/// The error type for [`new_widget`].
#[derive(thiserror::Error, Debug)]
pub enum NewLayerError {
    /// Snowcap returned a gRPC error status.
    #[error("gRPC error: `{0}`")]
    GrpcStatus(#[from] tonic::Status),
}

/// The error type for [`LayerHandle::update`] and set_* functions.
#[derive(thiserror::Error, Debug)]
pub enum UpdateLayerError {
    /// Snowcap returned a gRPC error status.
    #[error("gRPC error: `{0}`")]
    GrpcStatus(#[from] tonic::Status),
}

/// Create a new widget.
pub fn new_widget<Msg, P>(
    mut program: P,
    anchor: Option<Anchor>,
    keyboard_interactivity: KeyboardInteractivity,
    exclusive_zone: ExclusiveZone,
    layer: ZLayer,
) -> Result<LayerHandle<Msg>, NewLayerError>
where
    Msg: Clone + Send + 'static,
    P: Program<Message = Msg> + Send + 'static,
{
    let mut callbacks = HashMap::<WidgetId, WidgetMessage<Msg>>::new();

    let widget_def = program.view();

    widget_def.collect_messages(&mut callbacks, WidgetDef::message_collector);

    let response = Client::layer()
        .new_layer(NewLayerRequest {
            widget_def: Some(widget_def.clone().into()),
            anchor: anchor
                .map(From::from)
                .unwrap_or(layer::v1::Anchor::Unspecified) as i32,
            keyboard_interactivity: layer::v1::KeyboardInteractivity::from(keyboard_interactivity)
                as i32,
            exclusive_zone: exclusive_zone.into(),
            layer: layer::v1::Layer::from(layer) as i32,
        })
        .block_on_tokio()?;

    let layer_id = response.into_inner().layer_id;

    let mut event_stream = Client::widget()
        .get_widget_events(GetWidgetEventsRequest {
            id: Some(get_widget_events_request::Id::LayerId(layer_id)),
        })
        .block_on_tokio()?
        .into_inner();

    let (msg_send, mut msg_recv) = tokio::sync::mpsc::unbounded_channel::<Option<Msg>>();

    let handle = LayerHandle {
        id: layer_id.into(),
        msg_sender: msg_send.clone(),
    };

    if let Some(signaler) = program.signaler() {
        signaler.connect({
            let msg_send = msg_send.clone();

            move |msg: signal::Message<Msg>| {
                if let Err(err) = msg_send.send(Some(msg.into_inner())) {
                    error!("Failed to send emitted msg: {err}");
                    crate::signal::HandlerPolicy::Discard
                } else {
                    crate::signal::HandlerPolicy::Keep
                }
            }
        });

        signaler.connect({
            let msg_send = msg_send.clone();

            move |_: signal::RedrawNeeded| {
                if let Err(err) = msg_send.send(None) {
                    error!("Failed to send redraw signal: {err}");
                    crate::signal::HandlerPolicy::Discard
                } else {
                    crate::signal::HandlerPolicy::Keep
                }
            }
        });

        signaler.connect({
            let handle = handle.clone();

            move |operation: operation::Operation| {
                handle.operate(operation);
                crate::signal::HandlerPolicy::Keep
            }
        });
    }

    program.created(handle.clone().into());

    tokio::spawn(async move {
        loop {
            tokio::select! {
                Some(Ok(response)) = event_stream.next() => {
                    for widget_event in response.widget_events {
                        let Some(msg) = widget::message_from_event(&callbacks, widget_event) else {
                            continue;
                        };

                        program.update(msg);
                    }
                }
                Some(msg) = msg_recv.recv() => {
                    if let Some(msg) = msg {
                        program.update(msg);
                    }

                    if let Err(status) = Client::layer()
                        .request_view(ViewRequest { layer_id })
                        .block_on_tokio()
                    {
                        error!("Failed to request view for {layer_id}: {status}");
                    }

                    continue;
                }
                else => break,
            };

            let widget_def = program.view();

            callbacks.clear();

            widget_def.collect_messages(&mut callbacks, WidgetDef::message_collector);

            Client::layer()
                .update_layer(UpdateLayerRequest {
                    layer_id,
                    widget_def: Some(widget_def.into()),
                    anchor: None,
                    keyboard_interactivity: None,
                    exclusive_zone: None,
                    layer: None,
                })
                .await
                .unwrap();
        }
    });

    Ok(handle)
}

/// A handle to a layer surface.
#[derive(Clone)]
pub struct LayerHandle<Msg> {
    id: WidgetId,
    msg_sender: UnboundedSender<Option<Msg>>,
}

impl<Msg> std::fmt::Debug for LayerHandle<Msg> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("LayerHandle").field("id", &self.id).finish()
    }
}

impl<Msg> LayerHandle<Msg> {
    /// Update this layer's attributes.
    pub fn update(
        &self,
        anchor: Option<Option<Anchor>>,
        keyboard_interactivity: Option<KeyboardInteractivity>,
        exclusive_zone: Option<ExclusiveZone>,
        layer: Option<ZLayer>,
    ) -> Result<(), UpdateLayerError> {
        let anchor = if let Some(anchor) = anchor {
            anchor
                .map(layer::v1::Anchor::from)
                .or(Some(layer::v1::Anchor::Unspecified))
                .map(i32::from)
        } else {
            None
        };

        let keyboard_interactivity = keyboard_interactivity
            .map(layer::v1::KeyboardInteractivity::from)
            .map(i32::from);

        let exclusive_zone = exclusive_zone.map(i32::from);

        let layer = layer.map(layer::v1::Layer::from).map(i32::from);

        Client::layer()
            .update_layer(UpdateLayerRequest {
                layer_id: self.id.to_inner(),
                widget_def: None,
                anchor,
                keyboard_interactivity,
                exclusive_zone,
                layer,
            })
            .block_on_tokio()?;

        Ok(())
    }

    /// Update this layer's anchor.
    pub fn set_anchor(&self, anchor: Option<Anchor>) -> Result<(), UpdateLayerError> {
        self.update(Some(anchor), None, None, None)
    }

    /// Update this layer's keyboard_interactivity.
    pub fn set_keyboard_interactivity(
        &self,
        keyboard_interactivity: KeyboardInteractivity,
    ) -> Result<(), UpdateLayerError> {
        self.update(None, Some(keyboard_interactivity), None, None)
    }

    /// Update this layer's exclusive_one.
    pub fn set_exclusive_zone(
        &self,
        exclusive_zone: ExclusiveZone,
    ) -> Result<(), UpdateLayerError> {
        self.update(None, None, Some(exclusive_zone), None)
    }

    /// Update this layer's ZLayer.
    pub fn set_layer(&self, layer: ZLayer) -> Result<(), UpdateLayerError> {
        self.update(None, None, None, Some(layer))
    }

    /// Close this layer widget.
    pub fn close(&self) {
        if let Err(status) = Client::layer()
            .close(CloseRequest {
                layer_id: self.id.to_inner(),
            })
            .block_on_tokio()
        {
            error!("Failed to close {self:?}: {status}");
        }
    }

    /// Sends a message to this Layer [`Program`].
    pub fn send_message(&self, message: Msg) {
        let _ = self.msg_sender.send(Some(message));
    }

    /// Forces this layer to redraw.
    pub fn force_redraw(&self) {
        let _ = self.msg_sender.send(None);
    }

    /// Sends an [`Operation`] to this Layer.
    ///
    /// [`Operation`]: widget::operation::Operation
    pub fn operate(&self, operation: widget::operation::Operation) {
        if let Err(status) = Client::layer()
            .operate_layer(OperateLayerRequest {
                layer_id: self.id.to_inner(),
                operation: Some(operation.into()),
            })
            .block_on_tokio()
        {
            error!("Failed to send operation to {self:?}: {status}");
        }
    }
}

impl<Msg> LayerHandle<Msg>
where
    Msg: Clone + Send + 'static,
{
    /// Do something when a key event is received
    pub fn on_key_event(
        &self,
        mut on_event: impl FnMut(LayerHandle<Msg>, KeyEvent) + Send + 'static,
    ) {
        let mut stream = match Client::input()
            .keyboard_key(KeyboardKeyRequest {
                target: Some(Target::LayerId(self.id.to_inner())),
            })
            .block_on_tokio()
        {
            Ok(stream) => stream.into_inner(),
            Err(status) => {
                error!("Failed to set `on_key_event` handler: {status}");
                return;
            }
        };

        let handle = self.clone();

        tokio::spawn(async move {
            while let Some(Ok(response)) = stream.next().await {
                let event = KeyEvent::from(response);

                on_event(handle.clone(), event);
            }
        });
    }

    /// Do something on key press.
    pub fn on_key_press(
        &self,
        mut on_press: impl FnMut(LayerHandle<Msg>, Keysym, Modifiers) + Send + 'static,
    ) {
        self.on_key_event(move |handle, event| {
            if !event.pressed || event.captured {
                return;
            }

            on_press(handle, event.key, event.mods)
        });
    }
}

impl<Msg> AsParent for LayerHandle<Msg> {
    fn as_parent(&self) -> crate::popup::Parent {
        popup::Parent(popup::ParentInner::Layer(self.id))
    }
}
